src/sink/kafka.rs

```diff
@@ -0,0 +1,214 @@
+                match result {
+                    Ok(inner) => match inner {
+                        Ok((_partition, _offset)) => {
+                            KAFKA_PUBLISH_SUCCESS_SUM.fetch_add(1, Ordering::Relaxed);
+                            None
+                        }
+
+                        Err((err, message)) => match err {
+                            KafkaError::MessageProduction(err) => match err {
+                                RDKafkaError::InvalidMessage
+                                | RDKafkaError::UnknownTopicOrPartition
+                                | RDKafkaError::LeaderNotAvailable
+                                | RDKafkaError::NotLeaderForPartition
+                                | RDKafkaError::RequestTimedOut
+                                | RDKafkaError::NetworkException
+                                | RDKafkaError::GroupLoadInProgress
+                                | RDKafkaError::GroupCoordinatorNotAvailable
+                                | RDKafkaError::NotCoordinatorForGroup
+                                | RDKafkaError::NotEnoughReplicas
+                                | RDKafkaError::NotEnoughReplicasAfterAppend
+                                | RDKafkaError::NotController => {
+                                    KAFKA_PUBLISH_RETRY_SUM
+                                        .fetch_add(1, Ordering::Relaxed);
+                                    Some(message)
+                                }
+
+                                _ => {
+                                    error!("Kafka broker returned an unrecoverable error: {:?}", err);
+                                    KAFKA_PUBLISH_FAILURE_SUM
+                                        .fetch_add(1, Ordering::Relaxed);
+                                    None
+                                }
+                            },
+
+                            _ => {
+                                error!("Failed in send to kafka broker: {:?}", err);
+                                KAFKA_PUBLISH_FAILURE_SUM
+                                    .fetch_add(1, Ordering::Relaxed);
+                                None
+                            }
+                        },
+                    },
+
+                    _ => {
+                        error!("Failed in send to kafka broker: {:?}", result);
+                        KAFKA_PUBLISH_FAILURE_SUM.fetch_add(1, Ordering::Relaxed);
+                        None
+                    }
+                }
+            })
+            .collect();
+        let new_messages = retry_payload_and_keys
+            .iter()
+            .map(|maybe_retry| match *maybe_retry {
+                Some(ref message) => {
+                    let payload = message.payload();
+                    let key = message.key();
+                    if payload.is_some() && key.is_some() {
+                        Some(self.try_payload(payload.unwrap(), key.unwrap()))
```

Comment:
The unwrapping here makes me a touch nervous. Under what conditions would we hit the blow-up branch?

Reply:
Should be no blow up here. The unwraps are guarded by the is_some checks in the if. TBH, I'm not sure under what conditions these things will come back as None. I should update the else branch to add to a counter so we can track if it ever happens with the internal telemetry.

Reply:
+1 on the additional counter.